services/src/handlers/model_training.rs

```diff
@@ -0,0 +1,154 @@
+use futures::StreamExt;
+use geoengine_datatypes::primitives::MachineLearningQueryRectangle;
+
+#[cfg(feature = "xgboost")]
+use geoengine_datatypes::pro::MachineLearningFeature;
+use geoengine_datatypes::raster::GridOrEmpty;
+use geoengine_operators::call_on_generic_raster_processor;
+use geoengine_operators::engine::{
+    ExecutionContext, InitializedRasterOperator, QueryContext, QueryProcessor, RasterOperator,
+    TypedRasterQueryProcessor,
+};
+use num_traits::AsPrimitive;
+
+use crate::error::Error;
+use crate::error::Result;
+use crate::workflows::workflow::Workflow;
+
+const BATCH_SIZE: usize = 1_000;
+
+#[derive(Debug)]
+enum MachineLearningAccumKind {
+    XGBoost(Vec<f32>),
+}
+
+#[cfg(feature = "xgboost")]
+#[derive(Debug)]
+/// Used to gather the source data before further processing/shaping for xgboost
+struct MachineLearningAccum {
+    feature_name: String,
+    accum: MachineLearningAccumKind,
+}
+
+impl MachineLearningAccum {
+    fn new(feature_name: String) -> MachineLearningAccum {
+        MachineLearningAccum {
+            feature_name,
+            accum: MachineLearningAccumKind::XGBoost(Vec::new()),
+        }
+    }
+
+    fn update(&mut self, values: impl Iterator<Item = f32>) {
+        match self.accum {
+            MachineLearningAccumKind::XGBoost(ref mut accumulator) => {
+                for chunk in &itertools::Itertools::chunks(values, BATCH_SIZE) {
+                    accumulator.extend(chunk.filter(|x| x.is_finite()));
+                }
+            }
+        }
+    }
+
+    fn finish(&mut self) -> MachineLearningFeature {
+        match &self.accum {
+            MachineLearningAccumKind::XGBoost(data) => {
+                MachineLearningFeature::new(Some(self.feature_name.clone()), data.clone())
+            }
+        }
+    }
+}
+
+/// Build ML Features from the raw data and assign feature names.
+pub async fn accumulate_raster_data(
+    feature_names: Vec<Option<String>>,
+    input: Vec<TypedRasterQueryProcessor>,
+    query: MachineLearningQueryRectangle,
+    qry_ctx: &dyn QueryContext,
+) -> Result<Vec<Result<MachineLearningFeature>>> {
+    let mut feature_counter = -1;
+
```

Comment:
add a todo for adding a way to abort the query execution when the tasks is aborted